src/backend/distributed/metadata/metadata_cache.c

```diff
@@ -307,73 +310,97 @@ bool
 IsCitusTableType(Oid relationId, CitusTableType tableType)
 {
 	CitusTableCacheEntry *tableEntry = LookupCitusTableCacheEntry(relationId);
+	if (tableEntry)
+	{
+		return IsCitusTableTypeCacheEntry(tableEntry, tableType);
+	}
 
-	/* we are not interested in postgres tables */
-	if (tableEntry == NULL)
+	if (CitusHasBeenLoaded())
 	{
+		/*
+		 * Even if Citus has been loaded, we couldn't find (or build) a cache
+		 * entry for relation. Then it's definitely a postgres table.
+		 */
 		return false;
 	}
-	return IsCitusTableTypeInternal(tableEntry, tableType);
+
+	/*
+	 * If we are issuing ALTER EXTENSION citus UPDATE, CitusHasBeenLoaded()
+	 * returns false and we cannot access/build cache entries. In that case,
+	 * we need to decide table type directly reading from pg_dist_partition
+	 * table.
+	 * Reading from pg_dist_partition wouldn't be as efficient as accessing
+	 * to cache, but this is the case only when upgrading/downgrading Citus
+	 * so we don't expect any performance critical code-paths to reach here.
+	 */
+	ereport(DEBUG4, (errmsg("will scan pg_dist_partition for relation with "
```

Comment:
In most contexts, CitusHasBeenLoaded means the extension has been successfully created. 

Here when it returns false we still assume it is created and that we can read from the tables that are part of the extension. I'm not convinced those are always true here.

What is the code path that calls this? Should we prevent that code path from calling IsCitusTableType by checking CitusHasBeenLoaded() earlier?

Reply:
This udf call from other pr was hitting to this problem: https://github.com/citusdata/citus/pull/4174/files#diff-2c5ab3cbb878755ca18c565986d3e62fR12

In that pr, we definitely know that `CitusHasBeenLoaded()` returns false as we are executing migration scripts (so we weren't able to catch citus local tables there).

I could simply call the method that decides table type by reading from catalog table in that pr, but wanted to make this change in `IsCitusTableType` function to unify the logic.

Reply:
>Here when it returns false we still assume it is created and that we can read from the tables that are part of the extension. I'm not convinced those are always true.

I'm re-working on that part now, will update pr soon

Reply:
Instead of using that UDF, could we just query the catalog directly from the downgrade script?